src/lib.rs

```diff
@@ -90,27 +90,30 @@ impl Slice {
             // end to Item::Default (which arrives here as a None).
             end.map_or(Bound::Included(0), Bound::Excluded)
         };
-        RangeIterator::new((len - 1).min(start), end, step)
+        RangeIterator::new(start, end, step)
     }
 }
 
 impl Index {
     /// absolute index. negative indices are added to len.
     fn abs(&self, len: usize) -> Option<usize> {
         match self {
-            &Head(n) => Some(len.min(n)),
-            &Tail(n) => {
-                if len < n {
-                    None
-                } else {
-                    Some(len - n)
-                }
-            }
+            &Head(n) => ensure_within(0, n, len),
+            &Tail(n) => ensure_within(0, n, len).map(|n| len - n),
             Default => None,
         }
     }
 }
 
+/// Return Some(n) if min <= n < max, otherwise return None.
```

Comment:
I think having non-inclusive end range is more clear (and we can also use std::ops::Range)

Reply:
@glyn PTAL